lib/policy_machine_storage_adapters/active_record/postgresql.rb

```diff
@@ -13,34 +12,86 @@ def self.transitive_closure?(ancestor, descendant)
+          [*element_or_scope].map(&:id)
+        )
       end
 
       def self.ancestors_of(element_or_scope)
-        #FIXME: Also, removing the superfluous join of Assignment onto the recursive call is hugely beneficial to performance, but not supported
+        # FIXME: Removing the superfluous join of Assignment onto the recursive call is hugely beneficial to performance, but not supported
         # by hierarchical_query. Since this is a major performance pain point, generating raw SQL for now.
-        element_or_scope = [*element_or_scope]
-        case element_or_scope.size
-        when 0
-          PolicyElement.none
-        when 1
-          PolicyElement.where('"policy_elements"."id" IN (SELECT assignments__recursive.parent_id FROM (WITH RECURSIVE "assignments__recursive" AS ( SELECT "assignments"."id", "assignments"."parent_id", "assignments"."child_id" FROM "assignments" WHERE "assignments"."child_id" = ? UNION ALL SELECT "assignments"."id", "assignments"."parent_id", "assignments"."child_id" FROM "assignments" INNER JOIN "assignments__recursive" ON "assignments__recursive"."parent_id" = "assignments"."child_id" ) SELECT "assignments__recursive".* FROM "assignments__recursive") AS "assignments__recursive")', element_or_scope.first.id)
-        else
-          PolicyElement.where('"policy_elements"."id" IN (SELECT assignments__recursive.parent_id FROM (WITH RECURSIVE "assignments__recursive" AS ( SELECT "assignments"."id", "assignments"."parent_id", "assignments"."child_id" FROM "assignments" WHERE "assignments"."child_id" in (?) UNION ALL SELECT "assignments"."id", "assignments"."parent_id", "assignments"."child_id" FROM "assignments" INNER JOIN "assignments__recursive" ON "assignments__recursive"."parent_id" = "assignments"."child_id" ) SELECT "assignments__recursive".* FROM "assignments__recursive") AS "assignments__recursive")', element_or_scope.map(&:id))
-        end
+        PolicyElement.where(
+          'id IN (
+            WITH RECURSIVE assignments_recursive AS (
+              (
+                SELECT parent_id, child_id
+                FROM assignments
+                WHERE child_id IN (?)
+              )
+              UNION ALL
+              (
+                SELECT assignments.parent_id, assignments.child_id
+                FROM assignments
+                INNER JOIN assignments_recursive
+                ON assignments_recursive.parent_id = assignments.child_id
+              )
+            )
+
+            SELECT assignments_recursive.parent_id
+            FROM assignments_recursive
+          )',
+          [*element_or_scope].map(&:id)
+        )
       end
 
+      # Returns the operation set IDs from the given list where the operation is
+      # a descendant of the operation set.
+      # TODO: Generalize this so that we can arbitrarily filter recursive assignments calls.
+      def self.filter_operation_set_list_by_assigned_operation(operation_set_ids, operation_id)
+        query = <<-SQL
```

Comment:
style seems inconsistent, above using long strings in single quotes.  Still the mild preference for heredocs, but consistent seems gooder.